scripts/upgrade

```diff
@@ -1,141 +1,227 @@
+	ynh_app_setting_delete --app=$app --key=is_public
 fi
 
+# MIGRATION: Remove old code (from pre-4.1.5 versions, not using venv)
+if [[ "$upgrade_from_opt" == "true" ]]; then
 
+	# Remove legacy install dir
+	ynh_secure_remove /opt/yunohost/ihatemoney
 
-# MIGRATION: minor Py version has changed ? rebuilt venv
+	# Remove legacy Supervisor config
+	rm -f /etc/supervisor/conf.d/ihatemoney.conf
 
-# Useful for  Py 3.4 → 3.5, Jessie → Stretch, ynh 2.x → 3.x
-if [[ "$VENV_PY_VERSION" != '2.7' ]] && [[ "$VENV_PY_VERSION" != "$SYSTEM_PY_VERSION" ]]
-then
-    mv ${INSTALL_DIR}/venv ${INSTALL_DIR}/venv-old
-    init_virtualenv
-
-    # the static path changed
-    configure_nginx "$domain" "$path"
-fi
-
-#-------------------------------UPGRADE-------------------------
-
-
-# Upgrade code and dependencies
-pip_install
+	if [ -e /etc/ihatemoney/settings.py ]; then
+		# Strip out the no longer used part of the settings
+		python3 -c "d = open('/etc/ihatemoney/settings.py').read().replace('try:\n    from settings import *\nexcept ImportError:\n    pass\n', ''); open('/etc/ihatemoney/settings.py', 'w').write(d)"
+		# Rename
+		mv /etc/ihatemoney/settings.py "/etc/ihatemoney/ihatemoney.cfg"
+	fi
 
+	for old_file in "/etc/ihatemoney/ihatemoney.cfg" "/etc/$app/gunicorn.conf.py"; do
+		ynh_backup_if_checksum_is_different --file="$old_file"
+		ynh_delete_file_checksum --file="$old_file"
+	done
 
-#-----------------------POST-UPGRADE MIGRATIONS-----------------
-
+	final_path=/var/www/$app
+	ynh_app_setting_set --app=$app --key=final_path --value=$final_path
 
+	db_name=ihatemoney
+	db_user=$db_name
+	ynh_app_setting_set --app=$app --key=db_name --value=$db_name
+fi
 
+#=================================================
+# CREATE DEDICATED USER
+#=================================================
+ynh_script_progression --message="Making sure dedicated system user exists..." --weight=1
 
-# Python-MySQL is no longer maintained and does not support Py3
-ynh_replace_string "'mysql://" "'mysql+pymysql://" ${ihatemoney_conf_path}
+# Create a dedicated user (if not existing)
+ynh_system_user_create --username=$app --home_dir="$final_path"
+
+#=================================================
+# UPGRADE DEPENDENCIES
+#=================================================
+ynh_script_progression --message="Upgrading dependencies..." --weight=1
 
+ynh_install_app_dependencies $pkg_dependencies
+
+#=================================================
+# SPECIFIC UPGRADE
+#=================================================
+# Init venv
+#=================================================
+ynh_script_progression --message="Configuring the app's installation..." --weight=6
 
+ynh_app_setting_set --app=$app --key=final_path --value=$final_path
 
-# MIGRATION: Remove old code (from pre-2.x versions, not using pip)
+# MIGRATION: Upgrade venv
+python3 -m venv --upgrade "$final_path/venv"
+"$final_path/venv/bin/python3" -m pip install --upgrade pip setuptools "${pip_dependencies[@]}"
```

Comment:
```suggestion
"$final_path/venv/bin/python3" -m pip install --upgrade pip "${pip_dependencies[@]}"
```

(otherwise, setuptools is mentioned twice and pip issues a warning)

Reply:
@Salamandar suprisingly, applying that patch made the difference for me (I guess pip returns nonzero when issuing this warning) and made the upgrade pass !